### DISCUSSION TITLE
comparison with nx

### DISCUSSION THREAD

**Discussion by threepointone** [jpmorganchase/modular#375] [General]
[nx](nx.dev) is a feature-heavy toolchain for managing monorepos. It's gaining a lot of acceptance; and because they're building a business around it, is also very well supported. This issue isn't meant to be a deep dive into every nx feature, but only some of the key ideas in the context of modular. 

---

[e2e](https://nx.dev/latest/react/cli/e2e): Standardises on cypress for writing and running e2e tests. This is a good idea, we should do this too. (related - https://github.com/jpmorganchase/modular/issues/48)

[dep-graph](https://nx.dev/latest/react/cli/dep-graph): People like this, though it's not clear to me just yet what "business value" it brings. Regardless, it should be something we consider for "modular dashboard" (or whatever we call it.)

[affected](https://nx.dev/latest/react/cli/affected): This is one of the big features of nx, being able to run builds/lint/test/whatever only on workspaces/modules that have been changes/affected by other changed files. We should definitely work on this (related https://github.com/jpmorganchase/modular/issues/147, but we should expand the scope more)

[computation caching](https://nx.dev/latest/react/workspace/computation-caching): related to `affected`. (related issue: https://github.com/jpmorganchase/modular/issues/121)

[console](https://nx.dev/latest/react/cli/console): This is dope, and we should look to it when building modular-dashboard. 

[migrate](https://nx.dev/latest/react/cli/migrate): Interesting story on upgrades (for nx itself, iiuc). I still firmly believe that we can keep breaking changes to a minimum post-1.0 so it won't be a _major_ concern, but having automated workflows for upgrades sounds very appealing (much like codemods for files)

[plugins](https://nx.dev/latest/react/plugins/overview): Might not be something we need/should do, considering we're trying to be as opinionated as possible. 

---

(This was a quick run through, I'll probably keep updating this list and adding more detail as it comes to light.)

**Comment by matt-d-webb** [comment]
> console: This is dope, and we should look to it when building modular-dashboard.

Wow, that looks pretty neat. I think they've touched on the correct point here when they say Developers _"use the right tool for the job"_ (or at least try to). Not everything makes sense (or is easy) via the CLI. We should certainly make the correct distinction with our toolset too, having a VS Code plugin has a lot of potential to bring together a number of highly valuable features!
